pkg/upgrade/control_plane.go

```diff
@@ -733,21 +786,29 @@ func (u *ControlPlaneUpgrader) updateInfrastructureReference(replacementKey ctrl
 	// Step 2: if we're here, we need to create it
 
 	// get original infrastructure object
-	infraRef, err := external.Get(u.managementClusterClient, &ref, u.clusterNamespace)
+	infra, err := external.Get(u.managementClusterClient, &ref, u.clusterNamespace)
 	if err != nil {
 		return err
 	}
 
 	// prep the replacement
-	infraRef.SetResourceVersion("")
-	infraRef.SetName(replacementKey.Name)
-	infraRef.SetOwnerReferences(nil)
-	unstructured.RemoveNestedField(infraRef.UnstructuredContent(), "spec", "providerID")
+	infra.SetResourceVersion("")
+	infra.SetName(replacementKey.Name)
+	infra.SetOwnerReferences(nil)
+	unstructured.RemoveNestedField(infra.UnstructuredContent(), "spec", "providerID")
+
+	// Convert to a runtime.Object in case we need to patch, so we don't have to type assert after patching
+	var toCreate runtime.Object = infra
 
-	// point the machine at the replacement
+	if u.infrastructurePatch != nil {
+		toCreate, err = patchRuntimeObject(infra, u.infrastructurePatch)
+		if err != nil {
+			return errors.Wrap(err, "error patching infrastructure resource")
+		}
+	}
 
 	// create the replacement infrastructure object
-	err = u.managementClusterClient.Create(context.TODO(), infraRef)
+	err = u.managementClusterClient.Create(context.TODO(), toCreate)
```

Comment:
weird. I got a strange stack trace here:

```
INFO[0000] Updating infrastructure reference             api-version=infrastructure.cluster.x-k8s.io/v1alpha2 kind=DockerMachine machine=default/controlplane-0 name=controlplane-0 upgrade-id=1
Error: the server could not find the requested resource
the server could not find the requested resource
github.com/vmware/cluster-api-upgrade-tool/pkg/upgrade.(*ControlPlaneUpgrader).updateInfrastructureReference
	/home/ubuntu/cluster-api-upgrade-tool/pkg/upgrade/control_plane.go:813
github.com/vmware/cluster-api-upgrade-tool/pkg/upgrade.(*ControlPlaneUpgrader).updateMachines
	/home/ubuntu/cluster-api-upgrade-tool/pkg/upgrade/control_plane.go:571
github.com/vmware/cluster-api-upgrade-tool/pkg/upgrade.(*ControlPlaneUpgrader).Upgrade
	/home/ubuntu/cluster-api-upgrade-tool/pkg/upgrade/control_plane.go:216
main.main.func1
	/home/ubuntu/cluster-api-upgrade-tool/main.go:56
github.com/spf13/cobra.(*Command).execute
	/home/ubuntu/go/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:762
github.com/spf13/cobra.(*Command).ExecuteC
	/home/ubuntu/go/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:852
github.com/spf13/cobra.(*Command).Execute
	/home/ubuntu/go/pkg/mod/github.com/spf13/cobra@v0.0.3/command.go:800
main.main
	/home/ubuntu/cluster-api-upgrade-tool/main.go:118
runtime.main
	/home/ubuntu/.gimme/versions/go1.12.12.linux.amd64/src/runtime/proc.go:200
runtime.goexit
	/home/ubuntu/.gimme/versions/go1.12.12.linux.amd64/src/runtime/asm_amd64.s:1337
```

I would expect "could not create" but not "could not find" 🤔

Reply:
I think this means it couldn't find the `dockermachine` CRD type. Maybe a bug in this code somewhere?

Reply:
Oh I wonder if we lost the typemeta

Reply:
definitely lost typemeta.